### DISCUSSION TITLE
Bulk editing for materials?

### DISCUSSION THREAD

**Discussion by dibli-goost** [bnpr/Malt#351] [Feedback & Feature Requests]
One of my biggest frustrations with Malt right now is a lack of bulk editing for certain functions in Malt. Primarily for the light group settings and new property values on the main/pre pass outputs. _(Unless if there is a way to edit them for all materials that I'm not aware of.)_

If I want to change the light groups for all materials in a project, I have to manually go through and change each an every one of them. Which I know, it's a good idea to set them when you first create the material, but it's hard to experiment with them if I want to try new stuff out, and inevitably will lead to me not touching them if I have to edit 50 plus materials any time I want to mess with them.

If I add a new property to the main pass, for use in a screen pass for example, that property will get a default value for _every_ material in the project, which is fine! Until that default value is not preferable for whatever process I'm gonna be performing. Then I have to again, edit every single material.

---------------------------------

I'm not really sure how to implement bulk editing properly, and I'm sure that this is an issue with other aspects of the engine too, but I feel like it's a necessary workflow addition, it'll make managing large quantities of materials much more... manageable!

Maybe there could be an F3 searchable command for the light groups, set all material light/shadow groups to (X, X, X, X)?

**Comment by pragma37** [comment]
For overriding all Material light groups I think maybe an easy way to implement that would be to do it inside a node in the Render node tree.
One of the (not yet implemented) goals of Render nodes is making scene modifications before rendering.

For modifying all custom outputs for every node tree at the same time I can't think of a good solution.
Modifying the actual value (as in, setting a constant color) would be doable, but I doubt that's enough?

**Reply by dibli-goost** [reply]
While being able to do it in the Render nodes would be useful too, I think in my case I'll actually need to be change the values for each material. 

I'm currently using malt for a long form comic production, and am in the process of creating the assets.. with my shader set up (so far) I only really need two light groups, one for characters one for backgrounds.

The goal is of course to link characters and backgrounds to be used in new scenes. Each individual asset file only really contains materials of one light group type, so being able to quickly change the groups in that file would be preferable. Otherwise having to spend an hour changing all the light groups for the 20+ characters in the story would get rather tiresome.

Overrides would only really be feasible in situations where I needed just 1 light group in the scene I think. At least, I don't see a way for being able to tell Malt to selectively change the groups based on where the asset is linked from?

Something that I don't see a way around here is light groups that are decided upon inside the node trees, like on the various shader nodes. I don't think there's gonna be a way to bulk edit them without changing how they function. Maybe a suggestion for #348 here: what about linking those internal light group values to the material light groups by default? Which you can then override in the node tree if you need to? It could break a lot of people's projects but it'd be better to get a sweeping change like this in early.

-------------------

In regards to custom outputs. I haven't had a situation where I needed to bulk add nodes to node trees to input into one of them. (Idk how else to phrase this, hopefully you get it) I've only needed to change the actual value so far, most of the time I'm adding outputs I only need them for select few materials. But that's just me.

- addendum so that I don't forget: do we have a way to mask parts of a render out via what light group values for use in the Render nodes? If not, maybe that'd be a good feature to add. I'll make a proper feature request if not.

**Reply by pragma37** [reply]
> At least, I don't see a way for being able to tell Malt to selectively change the groups based on where the asset is linked from?

It would be doable, but I think the best solution would be filtering by Collection.

> Something that I don't see a way around here is light groups that are decided upon inside the node trees,

That's where Node Groups would be handy.

> what about linking those internal light group values to the material light groups by default? 

That's what the nodes in NPR Pipeline - NPR Mesh do.
But yes, the plan for the node redesign is to merge the nodes in NPR Mesh and NPR Shading and (in the case of Mesh shaders) use the material parameters by default.

>  I've only needed to change the actual value so far

I could look into supporting that, but it would be after we complete the node redesign.

> do we have a way to mask parts of a render out via what light group values for use in the Render nodes?

There's no built-in node, but render nodes can access that info.
You can also read the values from the material itself and send them to a custom output.
They're not directly available from nodes, but I guess they should be added.

You could use the Inline Code node for now:
![imagen](https://user-images.githubusercontent.com/36610999/178771602-75198448-d30f-4b66-ab5f-16e2f92d0ab5.png)

Or add the node to your custom library:
``` glsl
#ifdef MESH_SHADER
ivec4 get_light_groups(){ return MATERIAL_LIGHT_GROUPS; }
#endif
```
